File: cypress/integration/index/cookie_banner.spec.js
describe('Cookie banner', () => {
  before(() => {
    cy.visit('/')
    cy.checkEnvBanner()
  })

  const bannerMessageElement = '.govuk-cookie-banner__content > p'

  context('not logged in and javascript enabled', () => {
    beforeEach(() => {
      cy.visit('/')
    })

    const changeCookingSettings = 'Change your cookie settings'
    const rejectedAdditionalCookies = "You've rejected additional cookies"
    const hideMessageText = 'Hide this message'

    it('can reject cookie preferences', () => {
      cy.getCookie('cookies_preferences_set').should('not.exist')
      cy.checkCookieValue('analytics_cookies_set', 'false')
      cy.get("[data-cy='reject_cookies']")
        .should('contain', 'Reject analytics cookies')
        .and('have.attr', 'href')
        .and('include', 'analytics_cookies_set=false&show_confirm_banner=true')
      cy.get("[data-cy='reject_cookies']").click()
      cy.runtimeA11yCheck(null, null, cy.a11yLog)
      cy.get('.govuk-cookie-banner__content > p').should(
        'contain',
        rejectedAdditionalCookies
      )
      cy.checkCookieValue('cookies_preferences_set', 'true')
      cy.checkCookieValue('analytics_cookies_set', 'false')
    })

    it('can accept cookie preferences', () => {
      cy.getCookie('cookies_preferences_set').should('not.exist')
      cy.checkCookieValue('analytics_cookies_set', 'false')
      cy.get("[data-cy='accept_cookies']")
        .should('contain', 'Accept analytics cookies')
        .and('have.attr', 'href')
        .and('include', 'analytics_cookies_set=true&show_confirm_banner=true')
      cy.get("[data-cy='accept_cookies']").click()
      cy.runtimeA11yCheck(null, null, cy.a11yLog)
      cy.get('.govuk-cookie-banner__content > p').should(
        'contain',
        "You've accepted additional cookies."
      )
      cy.checkCookieValue('cookies_preferences_set', 'true')
      cy.checkCookieValue('analytics_cookies_set', 'true')
    })

    context('cookies accepted', () => {
      beforeEach(() => {
        cy.get("[data-cy='accept_cookies']")
          .should('contain', 'Accept analytics cookies')
          .and('have.attr', 'href')
          .and(
            'include',
            'analytics_cookies_set=true&show_confirm_banner=true'
          )
        cy.get("[data-cy='accept_cookies']").click()
      })

      it('can hide confirmation message', () => {
        cy.get(bannerMessageElement).should(
          'contain',
          "You've accepted additional cookies"
        )
        cy.get("[data-cy='hide_message']")
          .should('contain', hideMessageText)
          .and('have.attr', 'href')
          .and('include', '?')

        cy.get("[data-cy='hide_message']").click()
        cy.runtimeA11yCheck(null, null, cy.a11yLog)
        cy.get("[data-cy='hide_message']").should('not.exist')
      })

      it('can go to cookie settings from cookie banner', () => {
        cy.get(`${bannerMessageElement} > a`)
          .contains(changeCookingSettings, { matchCase: false })
          .and('have.attr', 'href')
          .and('include', '/cookies/settings')
        cy.get(`${bannerMessageElement} > a`).click()
        cy.get('.govuk-heading-xl').should('contain', changeCookingSettings)
      })
    })

    context('cookies rejected', () => {
      beforeEach(() => {
        cy.get("[data-cy='reject_cookies']")
          .should('contain', 'Reject analytics cookies')
          .and('have.attr', 'href')
          .and(
            'include',
            'analytics_cookies_set=false&show_confirm_banner=true'
          )

        cy.get("[data-cy='reject_cookies']").click()
      })

      it('can hide confirmation message', () => {
        cy.get(bannerMessageElement).should(
          'contain',
          rejectedAdditionalCookies
        )
        cy.get("[data-cy='hide_message']")
          .should('contain', hideMessageText)
          .and('have.attr', 'href')
          .and('include', '?')

        cy.get("[data-cy='hide_message']").click()
        cy.runtimeA11yCheck(null, null, cy.a11yLog)
        cy.get("[data-cy='hide_message']").should('not.exist')
      })

      it('can go to cookie settings from cookie banner', () => {
        cy.get(`${bannerMessageElement} > a`)
          .contains(changeCookingSettings, { matchCase: false })
          .and('have.attr', 'href')
          .and('include', '/cookies/settings')
        cy.get(`${bannerMessageElement} > a`).click()
        cy.get('.govuk-heading-xl').should('contain', changeCookingSettings)
      })
    })

    it('can go to cookie settings', () => {
      cy.get("[data-cy='cookie_settings']")
        .should('contain', 'Cookie settings')
        .and('have.attr', 'href')
        .and('include', '/cookies/settings')

      cy.get("[data-cy='cookie_settings']").click()
      cy.get('.govuk-heading-xl').should('contain', changeCookingSettings)
    })
  })
})

File: spec/features/cookie_banner_spec.rb
# frozen_string_literal: true

RSpec.feature 'Cookies', type: :feature do
  context 'when cookies are accepted via banner' do
    before do
      visit '/'
      click_link_or_button 'Accept analytics cookies'
    end

    scenario 'confirmation banner is visible' do
      within '.app-cookie-banner' do
        expect(page).to have_text "You've accepted additional cookies."
        expect(page).to have_no_text 'Cookies on View Court Data'
      end
    end

    scenario 'confirmation message can link to cookie settings' do
      click_link_or_button 'change your cookie settings'

      expect(page).to have_current_path cookies_settings_path, ignore_query: true
      within '.govuk-main-wrapper' do
        expect(page).to have_css('.govuk-heading-xl', text: 'Change your cookie settings')
      end
    end

    scenario 'confirmation message can be hidden' do
      click_link_or_button 'Hide this message'

      expect(page).to have_no_css '.app-cookie-banner'
      expect(page).to have_no_text "You've accepted additional cookies"
    end

    scenario 'cookies setting form shows cookies are on' do
      click_link_or_button 'change your cookie settings'

      within '#new_cookie' do
        expect(find_by_id('cookie-analytics-true-field').checked?).to be true
      end
    end
  end

  context 'when cookies are rejected via banner' do
    before do
      visit '/'
      click_link_or_button 'Reject analytics cookies'
    end

    scenario 'confirmation banner is visible' do
      within '.app-cookie-banner' do
        expect(page).to have_text "You've rejected additional cookies."
        expect(page).to have_no_text('Cookies on View Court Data')
      end
    end

    scenario 'confirmation message can link to cookie settings' do
      click_link_or_button 'change your cookie settings'

      expect(page).to have_current_path cookies_settings_path, ignore_query: true
      within '.govuk-main-wrapper' do
        expect(page).to have_css('.govuk-heading-xl', text: 'Change your cookie settings')
      end
    end

    scenario 'confirmation message can be hidden' do
      click_link_or_button 'Hide this message'

      expect(page).to have_no_css '.app-cookie-banner'
      expect(page).to have_no_text "You've rejected additional cookies"
    end

    scenario 'cookies setting form shows cookies are off' do
      click_link_or_button 'change your cookie settings'

      within '#new_cookie' do
        expect(find_by_id('cookie-analytics-false-field').checked?).to be true
      end
    end
  end

  context 'when visting defendants page', :stub_defendants_uuid_urn_search do
    let(:user) { create(:user) }
    let(:case_urn) { 'TEST12345' }
    let(:defendant_id) { '844a6542-ffcb-4cd0-94ce-fda3ffc3081b' }

    before do
      allow(FeatureFlag).to receive(:enabled?).with(:defendants_page).and_return(true)
      sign_in user
      visit "laa_references/new?id=#{defendant_id}&urn=#{case_urn}"
      click_link_or_button 'Accept analytics cookies'
    end

    scenario 'confirmation banner is visible' do
      within '.app-cookie-banner' do
        expect(page).to have_text "You've accepted additional cookies."
        expect(page).to have_no_text 'Cookies on View Court Data'
      end
    end

    scenario 'confirmation message can be hidden' do
      click_link_or_button 'Hide this message'

      expect(page).to have_no_css '.app-cookie-banner'
      expect(page).to have_no_text "You've accepted additional cookies"
    end

    scenario 'cookies setting form shows cookies are on' do
      click_link_or_button 'change your cookie settings'

      within '#new_cookie' do
        expect(find_by_id('cookie-analytics-true-field').checked?).to be true
      end
    end
  end
end
